File: go/275.go
package main

func hIndex(citations []int) int {
	n := len(citations)
	l, r := 0, n-1
	if citations[r] == 0 {
		return 0
	}
	var mid int
	var t int
	var x int
	for l < r {
		mid = (l + r) >> 1
		t = citations[mid]
		x = n - mid
		if x <= t {
			r = mid
		} else {
			l = mid + 1
		}
	}
	return n - l
}

func main() {

}

File: python/275.py
from typing import List

class Solution:
    def hIndex(self, citations: List[int]) -> int:
        if citations[-1] == 0: return 0
        n = len(citations)
        l, r = 0, n - 1
        while l < r:
            mid = (l + r) // 2 + 1
            x = n - mid
            t = citations[mid]
            if t < x:
                r = mid - 1
            else:
                l = mid
        return min(n - l, citations[l])
